### ISSUE TITLE
support the ability to change the catalog location/version on a per user basis

### ISSUE THREAD

**Issue by jstrachan** [openshiftio/launchpad-addon#42]
to setup nice CI / CD pipelines of boosters on OSIO we're going to have to be able to test out different catalog versions (e.g. a PR for a catalog change) inside OSIO. 

So we'd like to be able do the following in a CI / CD Pipeline:

* for a given PR change to a particular booster version (e.g. when a new push is merged to a booster git repo), generate a new catalog called `foo` (git repo / branch / tarball / whatever) for just that single booster
* grab a test user account on OSIO and associate that user with the new `foo` booster catalog - which only contains the changed boosters (i.e. usually only 1)
* we then run the full E2E tests on OSIO for that user - testing all boosters in this new catalog (no need to test any other boosters)
* if the tests all pass the pipeline +1's the PR

This will ensure that all booster changes get tested on OSIO before they get committed into the OSIO booster catalog. (I suspect we may need different catalogs for different cluster kinds;  as some boosters will work on OSO Paid with large limits and not on OSO Free etc)

So we need a way to configure the forge addon so that it picks the catalog to use on a per user basis. The default mode for production users should be to use boosters baked into the docker image for speed & reliability; when we are running E2E tests its fine for things to be a bit more dynamic & slower (e.g. cloning the repo / downloading the tarball for the users custom catalog).

So the OSIO generator-backend/fabric8-generator will use the current user request to figure out if a custom catalog is required; then needs some way to pass that information into the launchpad-addon so it can use this custom catalog.

**Comment by jstrachan** [comment]
@gastaldi does the above sound OK to you? I can help get the E2E tests going so we can easily test this

**Comment by edewit** [comment]
further discussed here: https://github.com/openshiftio/appdev-planning/issues/44

**Comment by jstrachan** [comment]
yeah - we'll need the addon to be able to point at a custom catalog (e.g. from a PR, fork or whatever). I figured if we got this bit done first we could make more progress on  openshiftio/appdev-planning#44 - e.g. we could generate a catalog per booster and to CI / CD testing on each individual booster repo

**Comment by jstrachan** [comment]
this issue also avoids me trying to convince folks that we need many catalogs ;) - instead we can focus on using it as a mechanism of testing a booster before a new version gets merged into the global catalog (until the penny drops we're gonna need different catalogs for different platforms/clusters/users ;-)

**Comment by gastaldi** [comment]
That sounds interesting. How would that impact/what changes in the current structure that we have in the booster-catalog repository now?

**Comment by maxandersen** [comment]
> That sounds interesting. How would that impact/what changes in the current structure that we have in the booster-catalog repository now?

Don't think this would change anything on the current structure - this just asks to provide some api in the launch addon to allow where the list of boosters that will be available are being read from during tests.

**Comment by jstrachan** [comment]
it shouldn't have any affect at all on what structure / code you use to load/read/process a catalog; only that we need to be able to point the launchpad-addon at different catalogs for different users for testing (without having to do another OSIO release, bounce/modify pods or anything)

e.g. using the exact same [booster catalog](https://github.com/openshiftio/booster-catalog) right now, one strawman PR pipeline for a booster with full CI / CD on OSIO (so that PRs are validated on OSIO before merged into the main booster catalog) could be...
* create a branch on the [booster-catalog](https://github.com/openshiftio/booster-catalog) for the PR on the booster (e.g. PR-123-foo-1 for build 1 of PR 123 on booster foo)
* delete all other boosters from the catalog in the branch (to avoid testing every booster on every commit to ever booster)
* update the booster's git SHA to point to the PR/branch in the catalog branch - e.g. here https://github.com/openshiftio/booster-catalog/blob/master/crud/spring-boot/spring-boot-rest-http-crud-booster.yaml#L2
* use this new catalog branch/sha for a certain user in the OSIO forge server - while using the usual baked in catalog for regular users

Thats keeping the current git / sha approach to the catalog exactly as it is now - we'd just need to change the code in the addon to optionally check some request time parameter (maybe in the forge attributeMap?) to whether or not we should use the _global_ catalog as now or a _custom_ user catalog. 

Over time we could experiment with using a blob of YAML for the entire catalog with URLs to the source code tarballs (like Helm chart repos) - but this issue doesn't know or care what a catalog looks like! Do whatever I don't care ;)

In terms of the current implementation and code - if we could allow the git repo + branch/sha for the catalog to be specified on a per user basis in the launchpad-addon (e.g. via forge attribute map values) that'd work fine for OSIO. 

Note this only for testing; switching dynamically the catalog on a per request basis will probably be quite a bit slower than usual (as today we initialise the catalog on startup and it can take a few minutes) - though when testing we'd be usually testing out catalogs of only 1 or 2 boosters at a time so it should be fairly fast.

So if we had say, 2 attributes we could specify in the forge attributeMap for the github repo URL (or org name + repo name) and the git sha/branch for the catalog - that'd be enough for now. We can then configure those values on a per user basis in OSIO.

This would have the added benefit that folks could test/demo/try boosters in development/forks - on OSIO; or let folks add extra boosters themselves until they make their way into the global catalog

**Comment by gastaldi** [comment]
Right, that makes sense. FYI we index the catalog for some known branches during build time so that speeds up indexing a lot (https://github.com/openshiftio/launchpad-backend/blob/master/pom.xml#L174). I'll have a look at enhancing the addon to support such feature

**Comment by jstrachan** [comment]
Many thanks! As soon as you figure out how to pass those parameters I'll wire it into our backend and try figure out how to configure those values in the fabric8-ui

**Comment by jstrachan** [comment]
Once this is in place, maybe we could have a fork option in OSIO that forks the booster + booster catalog & points your own fork of the booster catalog at your own fork of the booster while you hack on your code & before you're ready to do a PR?

**Comment by gastaldi** [comment]
Yeah, that could work. A bit unrelated to this topic, but we should think of merging the backends at some point.

**Comment by jstrachan** [comment]
@gastaldi btw good to know on the build time indexing! :) 

For this use case - testing - it'd be no biggie if we index on the fly really. But I guess if the PR pipeline did the indexing at build time on the fork/branch & checked in the index to the catalog repo, your code could maybe reuse the index if its there - or just create a new index?

**Comment by jstrachan** [comment]
@gastaldi yeah - we had to fork the backend to add new forge commands anyway; Other than that its mostly a black box & shouldn't be too hard to reuse - though we've made some tweaks over time to the JAXRS bean - and I suspect we'll want to make lots more - as our entire wizards are all 100% pure Forge and many have many steps (sometimes 5-8 steps) so we had to do a few tweaks here or there to get it to work. 

We're hoping to move more of the wizard to the client side (so the client does a few stand alone GET commands to load github orgs, github repos, github account details, openshift namespaces, quickstart catalog and other stuff - which can be cached locally and polled for changes easily) then most of the wizard can be pure client side only to make things nice and snappy; then reuse the current multi-step wizard stuff just for the POST

**Comment by gastaldi** [comment]
That would be a good idea. Although you could have a better cache control in the serverside and return values as part of the same JSON, but that's just an impl detail.

**Comment by gastaldi** [comment]
About caching things, perhaps a custom CDI scope bound to the current logged in user would be a good solution

**Comment by jstrachan** [comment]
we've cached everything we can right now in our fabric8-generator on the server side; its often kinda complex as it very much depends on data from previous steps etc - but still the wizard is still pretty slow - (plus the first wizard load is super slow as the cache is all empty ;). 

Whereas most data rarely changes much (the github username/email, namespaces on openshift, the github orgs, the github repos, the quickstart names etc) - so the UI would be more responsive if we cached that data in localStorage so the UI appears instantly; then in the background polled/refreshed the UI in case there's recent changes.

So the move to the client side is more about faster initial load, page navigation & validation really; we'd prefer that to be a few background GETs then everything is mostly client side until folks hit `Finish`. We may need some custom validator POSTS in there here or there but its more an exception than the rule.

e.g. its just much easier to keep all this data on the client side all the time; so that as users type characters into, say, the project name field we can instantly respond with UX feedback if the project name is already in use in the users currently selected github organisation - rather than doing a full forge wizard validation REST call on each character press.

**Comment by gastaldi** [comment]
You can now specify `LAUNCHPAD_BACKEND_CATALOG_GIT_REPOSITORY` and `LAUNCHPAD_BACKEND_CATALOG_GIT_REF`as UIContext attributes and they will be used and properly cached when the command is invoked

**Comment by jstrachan** [comment]
awesome thanks!!! Which version of the addon do I need to use?

**Comment by gastaldi** [comment]
I'll release `io.openshift.launchpad:launchpad-addon:11` with these changes

**Comment by jstrachan** [comment]
@gastaldi btw is LAUNCHPAD_BACKEND_CATALOG_GIT_REPOSITORY the full github URL - like `https://github.com/openshiftio/booster-catalog.git` right?

**Comment by gastaldi** [comment]
Yes, that is correct

Em 17 de ago de 2017 12:59, "James Strachan" <notifications@github.com>
escreveu:

**Comment by gastaldi** [comment]
Let me know if that works for you

**Comment by jstrachan** [comment]
@gastaldi unfortunately we can't move to any new version of the launch addon any time soon in OSIO as it appears to be a complete rewrite from the version we've used in OSIO - so we'll have to do a massive rewrite of the OSIO forge addon to be able to deal with it - so it'll have to go to the back of my TODO list I'm afraid
